### DISCUSSION TITLE
PGN input option

### DISCUSSION THREAD

**Discussion by wwawrzy** [alexobviously/bishop#42] [Q&A] [answered]
I can't find an option to load a game from PGN. There is an option to set the FEN position. The description includes the phrase "FEN & PGN input and output". Is this option available somewhere?

Thank you for the great library!

**Comment by alexobviously** [comment]
Hi, actually PGN import is one of the things I'm currently working on, and since you need it, I'll prioritise it and get it in the next version.
Actually, there used to be an import function but it wasn't working very well as the scope of the library increased, but I always intended to replace it.

So yeah, maybe in the next few days I'll be overhauling the whole FEN/PGN system so it can accommodate custom logic better

**Comment by wwawrzy** [comment]
That's great. I'm currently working on a simple tool that allows me to analyze my chess game. To maintain compliance with the formal PGN notation, I plan to store special tags in the comments of individual moves. If you could provide the ability to also load comments, I would be grateful.

**Reply by wwawrzy** [reply]
I managed to write a simple test that goes through a list of PGN moves and generates FEN.

Below is the code and the execution time. Unfortunately, due to other commitments, I was not able to examine where the time difference is.

```
import 'package:bishop/bishop.dart' as bishop;
import 'package:chess/chess.dart';
import 'package:flutter/material.dart';

String pgn = 'f4 e5 e3 exf4 exf4 Nf6 Nf3 d5 d4 Bd6 Bd3 O-O O-O Nc6 c3 Ne7 Qc2 Ng6 Ne5 c5 Be3 c4 Be2 Ne4';
void main() {
  List<String> moves = pgn.split(' ');

  Stopwatch stopwatch = Stopwatch()..start();
  int fen_length = 0;
  for (var i = 0; i < 1000; i++) {
    var game = bishop.Game();
    for (var move in moves) {
      game.makeMoveSan(move);
      fen_length += game.fen.length;
    }
  }
  print('Time elapsed: ${stopwatch.elapsed} (${fen_length} characters)');

  Stopwatch stopwatch2 = Stopwatch()..start();
  fen_length = 0;
  for (var i = 0; i < 1000; i++) {
    Chess chess = new Chess();
    for (var move in moves) {
      chess.move(move);
      fen_length += chess.fen.length;
    }
  }
  stopwatch2.stop();
  print('Chess: Time elapsed: ${stopwatch2.elapsed} (${fen_length} characters)');
}
```

```
flutter: Time elapsed: 0:01:14.886033 (1521000 characters)
flutter: Chess: Time elapsed: 0:00:04.499805 (1521000 characters)
```

**Comment by alexobviously** [comment, marked as answer]
@wwawrzy hey, I added an optimisation in 1.2.5 - see my comment here (https://github.com/alexobviously/bishop/issues/44#issuecomment-1459652074) and see if it makes your script run faster.

**Reply by wwawrzy** [reply]
Thank you for the improvement. The performance is now very good.

**Reply by alexobviously** [reply]
Hey, you should try 1.3.1, there are some major performance improvements (attack checking much more optimised).
